quil-rs/src/stats/execution_graph.rs

```diff
@@ -0,0 +1,200 @@
+    ///
+    /// The `f` function is called for each instruction in each path, with the current accumulator value and the
+    /// current instruction.
+    ///
+    /// # Examples
+    ///
+    /// ## Tree
+    ///
+    /// ```text
+    /// CNOT 0 1
+    /// X 0
+    /// H 1
+    /// ```
+    ///
+    /// 1. `CNOT 0 1` is visited with the initial value, and a new accumulator `A` is returned from `f`.
+    /// 2. `X 0` is visited with accumulator `A`, and a result value `B` is returned from `f`.
+    /// 3. `H 1` is visited with accumulator `A`, and a second result value `C` is returned from `f`.
+    /// 4. The result values are collected into a [`Vec`] and returned as `[B, C]`.
+    ///
+    /// ## Diamond
+    ///
+    /// If the program graph forms a diamond shape (i.e. multiple paths converge to a single node), the `f` function
+    /// will be called multiple times with the same instruction, but with potentially different accumulator values.
+    ///
+    /// ```text
+    /// CNOT 0 1
+    /// X 0
+    /// H 1
+    /// CNOT 1 0
+    /// ```
+    ///
+    /// 1. `CNOT 0 1` is visited with the initial value, and a new accumulator `A` is returned from `f`.
+    /// 2. `X 0` is visited with accumulator `A`, and a new accumulator `B` is returned from `f`.
+    /// 3. `H 1` is visited with accumulator `A`, and a new accumulator `C` is returned from `f`.
+    /// 4. `CNOT 1 0` is visited with accumulator `B`, and a result value `D` is returned from `f`.
+    /// 5. `CNOT 1 0` is visited with accumulator `C`, and a result value `E` is returned from `f`.
+    /// 5. The result values are collected into a [`Vec`] and returned as `[D, E]`.
+    ///
+    /// # Errors
+    ///
+    /// Any error returned from a call to `f` will be returned immediately.
+    // TBD: visibility
+    fn path_fold<T, F>(&self, initial_value: T, mut f: F) -> Result<Vec<T>, Error>
+    where
+        T: Clone + std::fmt::Debug,
+        F: FnMut(T, &Instruction) -> Result<T, Error>,
+    {
+        let nodes: Vec<_> = self.graph.externals(Direction::Incoming).collect();
+        let mut stack = vec![(initial_value, nodes)];
+        let mut result = Vec::new();
+
+        while let Some((acc, nodes)) = stack.pop() {
+            if nodes.is_empty() {
+                result.push(acc);
+                continue;
+            }
+
+            for node in nodes {
+                let instruction = &self.graph[node];
+                let value = f(acc.clone(), instruction)?;
+                stack.push((
+                    value,
+                    self.graph
+                        .neighbors_directed(node, Direction::Outgoing)
+                        .collect(),
+                ));
+            }
+        }
+
+        Ok(result)
+    }
+
+    /// Returns the longest path from an initial instruction (one with no prerequisite instructions) to a final instruction (one with no dependent instructions).
+    pub fn gate_depth(&self) -> Result<usize, Error> {
```

Comment:
We had this return `Result` when we wanted to error on non-gate instructions. We now handle this better, so we can maybe make this infallible?

`path_fold` only returns a `Result` because the closures might return one. It looks like we are safe to make that infallible as well. If there's a good reason not to, we can also assert in this function that the result is `Result<_, Infallible>` so we get a compile error if we introduce a real error at some point.

Reply:
Here's a version where `path_fold` still takes a fallible callback, but `gate_depth` and `multi_qubit_gate_depth` use `Infallible`: https://github.com/rigetti/quil-rs/pull/336/commits/01d59cb36ecacd1516a26fbe12dd016a1d67a4b6#diff-73e235b2289330aa4f68043fbc217ba835e8adb3135e19c4a7b7e5f5a50b4479R103

I'll go ahead and just change the `path_fold` signature to require the callback not to return an error, though.

If we ever do have an execution graph visitor that can fail, we can just use `Result` as the `T` type, and then transpose from `Vec<Result<...>>` to `Result<Vec<...>, >` in the caller.

Reply:
...hm, since the callback takes `T` as an argument, that's actually not super convenient. In any case, I agree that there's probably no reason to have these things be fallible in the first place.